### PR TITLE
DLC Refactor/Rewrite

### PR DESCRIPTION
**Pull request by nkohen**
Refactored the entirety of what was formerly `DLCClient` into nice modules and used these modules in `DLCWallet` (for the wallet) and `TestDLCClient` (for tests)

TODO: Scaladocs

### CODE REVIEW

File: core/src/main/scala/org/bitcoins/core/wallet/builder/RawTxFinalizer.scala [hunk 1]
```diff
@@ -274,32 +275,69 @@ object StandardNonInteractiveFinalizer {
   }
 }
 
-case class SubtractFeeFromOutputsFinalizer(
-    inputInfos: Vector[InputInfo],
-    feeRate: FeeUnit)
+case class AddFutureFeeFinalizer(
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/wallet/builder/RawTxFinalizer.scala:282]: scaladoc

File: core/src/main/scala/org/bitcoins/core/wallet/builder/RawTxFinalizer.scala [hunk 2]
```diff
@@ -274,32 +275,69 @@ object StandardNonInteractiveFinalizer {
...
-    val dummyTxF = TxUtil.addDummySigs(txWithPossibleWitness, inputInfos)
+    val txT = outputsT.map { outputs =>
+      txBuilderResult.toBaseTransaction.copy(outputs = outputs)
+    }
+
+    Future.fromTry(txT)
+  }
+}
+
+case class SubtractFeeFromOutputsFinalizer(
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/wallet/builder/RawTxFinalizer.scala:329]: scaladoc

File: core/src/main/scala/org/bitcoins/core/wallet/builder/RawTxFinalizer.scala [hunk 3]
```diff
@@ -334,6 +370,78 @@ object SubtractFeeFromOutputsFinalizer {
       .dropRight(1)
       .:+((newLastOutput, lastOutputIndex))
 
-    (newOutputs ++ unchangedOutputs).sortBy(_._2).map(_._1).toVector
+    (newOutputs ++ unchangedOutputs).sortBy(_._2).map(_._1)
+  }
+}
+
+case class P2WPKHSubtractFeesFromOutputsFinalizer(
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/wallet/builder/RawTxFinalizer.scala:389]: scaladoc

File: core/src/main/scala/org/bitcoins/core/wallet/builder/RawTxFinalizer.scala [hunk 4]
```diff
@@ -334,6 +370,78 @@ object SubtractFeeFromOutputsFinalizer {
...
+    val outputsAfterFee =
+      SubtractFeeFromOutputsFinalizer.subtractFees(wtx, feeRate, spks)
+
+    Future.successful(
+      txBuilderResult.toBaseTransaction.copy(outputs = outputsAfterFee))
+  }
+}
+
+// Note: spendingVBytes may be 0
+case class DualFundingTxFinalizer(
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/wallet/builder/RawTxFinalizer.scala:418]: scaladoc

File: core/src/main/scala/org/bitcoins/core/wallet/builder/RawTxFinalizer.scala [hunk 5]
```diff
@@ -334,6 +370,78 @@ object SubtractFeeFromOutputsFinalizer {
...
+    val filterDust = FilterDustFinalizer
+
+    addFutureFee
+      .andThen(subtractFee)
+      .andThen(filterDust)
+      .buildTx(txBuilderResult)
+  }
+}
+
+case class P2WPKHDualFundingTxFinalizer(
```
Comment (nkohen) [inline core/src/main/scala/org/bitcoins/core/wallet/builder/RawTxFinalizer.scala:454]: scaladoc

File: dlc-test/src/test/scala/org/bitcoins/dlc/DLCClientIntegrationTest.scala [hunk 6]
```diff
@@ -30,6 +30,18 @@ import org.bitcoins.core.util.FutureUtil
...
+  CooperativeDLCOutcome,
+  DLCOutcome,
+  RefundDLCOutcome,
+  RefundDLCOutcomeWithClosing,
+  RefundDLCOutcomeWithDustClosing,
+  SetupDLC,
+  UnilateralDLCOutcome,
+  UnilateralDLCOutcomeWithClosing,
+  UnilateralDLCOutcomeWithDustClosing
+}
```
Comment (nkohen) [inline dlc-test/src/test/scala/org/bitcoins/dlc/DLCClientIntegrationTest.scala]: nit: `import org.bitcoins.dlc.execution._`

File: dlc-test/src/test/scala/org/bitcoins/dlc/DLCClientTest.scala [hunk 7]
```diff
@@ -19,11 +19,25 @@ import org.bitcoins.core.script.crypto.HashType
...
+  CooperativeDLCOutcome,
+  DLCOutcome,
+  RefundDLCOutcome,
+  RefundDLCOutcomeWithClosing,
+  RefundDLCOutcomeWithDustClosing,
+  SetupDLC,
+  UnilateralDLCOutcome,
+  UnilateralDLCOutcomeWithClosing,
+  UnilateralDLCOutcomeWithDustClosing
+}
```
Comment (nkohen) [inline dlc-test/src/test/scala/org/bitcoins/dlc/DLCClientTest.scala]: nit: `import org.bitcoins.dlc.execution._`

File: dlc/src/main/scala/org/bitcoins/dlc/builder/DLCCETBuilder.scala [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+  FilterDustFinalizer,
+  RawTxBuilder
+}
+import org.bitcoins.core.wallet.fee.FeeUnit
+import org.bitcoins.core.wallet.utxo.{ConditionalPath, P2WSHV0InputInfo}
+import org.bitcoins.crypto._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+case class DLCCETBuilder(
```
Comment (nkohen) [inline dlc/src/main/scala/org/bitcoins/dlc/builder/DLCCETBuilder.scala:24]: scaladoc

File: dlc/src/main/scala/org/bitcoins/dlc/builder/DLCCETBuilder.scala [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+      DLCTxBuilder.tweakedPubKey(fundingPubKey, toLocalCETKey, sigPubKeys(msg))
+
+    P2PKWithTimeoutScriptPubKey(
+      pubKey = pubKey,
+      lockTime = ScriptNumber(timeouts.penaltyTimeout.toLong),
+      timeoutPubKey = otherToLocalCETKey
+    )
+  }
+
+  def buildOfferToLocalP2PK(
```
Comment (nkohen) [inline dlc/src/main/scala/org/bitcoins/dlc/builder/DLCCETBuilder.scala:74]: scaladocs for all the methods

File: dlc/src/main/scala/org/bitcoins/dlc/builder/DLCFundingTxBuilder.scala [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+import org.bitcoins.core.wallet.builder.{
+  P2WPKHDualFundingTxFinalizer,
+  RawTxBuilderWithFinalizer
+}
+import org.bitcoins.core.wallet.fee.FeeUnit
+import org.bitcoins.crypto.ECPublicKey
+
+import scala.concurrent.{ExecutionContext, Future}
+
+case class DLCFundingTxBuilder(
```
Comment (nkohen) [inline dlc/src/main/scala/org/bitcoins/dlc/builder/DLCFundingTxBuilder.scala:23]: scaladoc for this and its public methods

File: dlc/src/main/scala/org/bitcoins/dlc/builder/DLCMutualCloseTxBuilder.scala [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+import org.bitcoins.commons.jsonmodels.dlc.DLCMessage.{ContractInfo, OracleInfo}
+import org.bitcoins.core.protocol.script.{EmptyScriptSignature, ScriptPubKey}
+import org.bitcoins.core.protocol.transaction._
+import org.bitcoins.core.wallet.builder.{FilterDustFinalizer, RawTxBuilder}
+import org.bitcoins.crypto.SchnorrDigitalSignature
+
+import scala.concurrent.{ExecutionContext, Future}
+
+case class DLCMutualCloseTxBuilder(
```
Comment (nkohen) [inline dlc/src/main/scala/org/bitcoins/dlc/builder/DLCMutualCloseTxBuilder.scala:12]: scaladoc for this and its public methods

File: dlc/src/main/scala/org/bitcoins/dlc/builder/DLCRefundTxBuilder.scala [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+  P2WPKHSubtractFeesFromOutputsFinalizer,
+  RawTxBuilder
+}
+import org.bitcoins.core.wallet.fee.FeeUnit
+import org.bitcoins.core.wallet.utxo.{ConditionalPath, P2WSHV0InputInfo}
+import org.bitcoins.crypto.ECPublicKey
+
+import scala.concurrent.{ExecutionContext, Future}
+
+case class DLCRefundTxBuilder(
```
Comment (nkohen) [inline dlc/src/main/scala/org/bitcoins/dlc/builder/DLCRefundTxBuilder.scala:24]: scaladoc for this and its public methods

File: dlc/src/main/scala/org/bitcoins/dlc/builder/DLCRefundTxBuilder.scala [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+    val offerValue = Satoshis(
+      (fundingOutput.value * offerInput).satoshis.toLong / totalInput.satoshis.toLong)
+    val acceptValue = Satoshis(
+      (fundingOutput.value * acceptInput).satoshis.toLong / totalInput.satoshis.toLong)
+
+    builder += TransactionOutput(offerValue, offerFinalSPK)
+    builder += TransactionOutput(acceptValue, acceptFinalSPK)
+
+    // TODO: Should be SubtractFeeFromOutputsFinalizer since we know fundingInputInfo
```
Comment (nkohen) [inline dlc/src/main/scala/org/bitcoins/dlc/builder/DLCRefundTxBuilder.scala:61]: make issue and fix in future PR (will require re-generating test vectors as fees will change)

File: dlc/src/main/scala/org/bitcoins/dlc/builder/DLCTxBuilder.scala [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+import org.bitcoins.core.number.UInt32
+import org.bitcoins.core.protocol.script.P2WSHWitnessV0
+import org.bitcoins.core.protocol.transaction._
+import org.bitcoins.core.protocol.{BitcoinAddress, BlockStampWithFuture}
+import org.bitcoins.core.wallet.fee.SatoshisPerVirtualByte
+import org.bitcoins.crypto._
+
+import scala.concurrent.{ExecutionContext, Future}
+
+case class DLCTxBuilder(
```
Comment (nkohen) [inline dlc/src/main/scala/org/bitcoins/dlc/builder/DLCTxBuilder.scala:19]: scaladoc

File: dlc/src/main/scala/org/bitcoins/dlc/builder/DLCTxBuilder.scala [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+  /** Returns the payouts for the signature as (toOffer, toAccept) */
+  def getPayouts(
+      oracleSig: SchnorrDigitalSignature): (CurrencyUnit, CurrencyUnit) = {
+    DLCTxBuilder.getPayouts(oracleSig,
+                            sigPubKeys,
+                            offerOutcomes,
+                            acceptOutcomes)
+  }
+
+  lazy val buildFundingTx: Future[Transaction] = {
```
Comment (nkohen) [inline dlc/src/main/scala/org/bitcoins/dlc/builder/DLCTxBuilder.scala:94]: scaladoc for this and methods below

File: dlc/src/main/scala/org/bitcoins/dlc/builder/DLCTxBuilder.scala [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+    sigPubKeys.find(_._2 == oracleSig.sig.getPublicKey) match {
+      case Some((hash, _)) =>
+        (offerOutcomes(hash), acceptOutcomes(hash))
+      case None =>
+        throw new IllegalArgumentException(
+          s"Signature does not correspond to a possible outcome! $oracleSig")
+    }
+  }
+
+  def tweakedPubKey(
```
Comment (nkohen) [inline dlc/src/main/scala/org/bitcoins/dlc/builder/DLCTxBuilder.scala:267]: scaladoc

File: dlc/src/main/scala/org/bitcoins/dlc/execution/DLCExecutor.scala [hunk 17]
```diff
@@ -0,0 +1,394 @@
...
+  P2WSHV0InputInfo,
+  ScriptSignatureParams
+}
+import org.bitcoins.crypto.{SchnorrDigitalSignature, Sha256DigestBE}
+import org.bitcoins.dlc.builder.DLCTxBuilder
+import org.bitcoins.dlc.sign.DLCTxSigner
+
+import scala.concurrent.{ExecutionContext, Future}
+
+case class DLCExecutor(signer: DLCTxSigner) extends BitcoinSLogger {
```
Comment (nkohen) [inline dlc/src/main/scala/org/bitcoins/dlc/execution/DLCExecutor.scala:33]: scaladoc for this and methods

File: dlc/src/main/scala/org/bitcoins/dlc/sign/DLCTxSigner.scala [hunk 18]
```diff
@@ -0,0 +1,325 @@
...
+  ECPublicKey,
+  FieldElement,
+  SchnorrDigitalSignature,
+  Sha256DigestBE
+}
+import org.bitcoins.dlc.builder.DLCTxBuilder
+
+import scala.concurrent.{ExecutionContext, Future}
+
+case class DLCTxSigner(
```
Comment (nkohen) [inline dlc/src/main/scala/org/bitcoins/dlc/sign/DLCTxSigner.scala:44]: scaladoc for this and methods

File: dlc/src/main/scala/org/bitcoins/dlc/sign/DLCTxSigner.scala [hunk 19]
```diff
@@ -0,0 +1,325 @@
...
+      fundingUtxos: Vector[ScriptSignatureParams[InputInfo]])(
+      implicit ec: ExecutionContext): DLCTxSigner = {
+    DLCTxSigner(DLCTxBuilder(offer, accept),
+                isInitiator,
+                extPrivKey,
+                nextAddressIndex,
+                fundingUtxos)
+  }
+
+  def tweakedPrivKey(
```
Comment (nkohen) [inline dlc/src/main/scala/org/bitcoins/dlc/sign/DLCTxSigner.scala:334]: scaladoc

File: dlc/src/main/scala/org/bitcoins/dlc/testgen/TestDLCClient.scala [hunk 20]
```diff
@@ -0,0 +1,313 @@
...
+  UnilateralDLCOutcome
+}
+import org.bitcoins.dlc.sign.DLCTxSigner
+import scodec.bits.ByteVector
+
+import scala.concurrent.duration.DurationInt
+import scala.concurrent.{Await, ExecutionContext, Future}
+
+/** This case class allows for the construction and execution of
+  * Discreet Log Contracts between two parties.
```
Comment (nkohen) [inline dlc/src/main/scala/org/bitcoins/dlc/testgen/TestDLCClient.scala]: between two parties running on this machine (for tests)

File: dlc/src/main/scala/org/bitcoins/dlc/testgen/TestDLCClient.scala [hunk 21]
```diff
@@ -0,0 +1,313 @@
...
+
+  lazy val fundingTxIdBE: DoubleSha256DigestBE = fundingTx.txIdBE
+
+  def createUnsignedMutualCloseTx(sig: SchnorrDigitalSignature): Transaction = {
+    val utxF = dlcTxBuilder.buildMutualCloseTx(sig)
+
+    Await.result(utxF, 5.seconds)
+  }
+
+  def setupDLCAccept(
```
Comment (nkohen) [inline dlc/src/main/scala/org/bitcoins/dlc/testgen/TestDLCClient.scala:88]: scaladocs for methods

File: dlc/src/main/scala/org/bitcoins/dlc/verify/DLCSignatureVerifier.scala [hunk 22]
```diff
@@ -0,0 +1,112 @@
...
+import org.bitcoins.core.psbt.InputPSBTRecord.PartialSignature
+import org.bitcoins.core.psbt.PSBT
+import org.bitcoins.crypto.Sha256DigestBE
+import org.bitcoins.dlc.builder.DLCTxBuilder
+
+import scala.concurrent.{Await, ExecutionContext}
+import scala.concurrent.duration.DurationInt
+import scala.util.{Failure, Success}
+
+case class DLCSignatureVerifier(builder: DLCTxBuilder, isInitiator: Boolean) {
```
Comment (nkohen) [inline dlc/src/main/scala/org/bitcoins/dlc/verify/DLCSignatureVerifier.scala:24]: scaladoc for this and methods

File: wallet/src/main/scala/org/bitcoins/wallet/DLCWallet.scala [hunk 23]
```diff
@@ -9,10 +9,19 @@ import org.bitcoins.core.number.UInt32
...
 import org.bitcoins.crypto._
-import org.bitcoins.dlc._
+import org.bitcoins.dlc.execution.{
+  DLCExecutor,
+  RefundDLCOutcomeWithClosing,
+  RefundDLCOutcomeWithDustClosing,
+  SetupDLC,
+  UnilateralDLCOutcomeWithClosing,
+  UnilateralDLCOutcomeWithDustClosing
+}
```
Comment (nkohen) [inline wallet/src/main/scala/org/bitcoins/wallet/DLCWallet.scala]: nit: `import org.bitcoins.dlc.execution._`

**Review by nkohen — COMMENTED** [review]
Marked where I need scaladocs and some other nits

File: wallet/src/main/scala/org/bitcoins/wallet/DLCWallet.scala [hunk 24]
```diff
@@ -497,121 +483,124 @@ abstract class DLCWallet extends Wallet {
...
+      fundingInputs: Vector[DLCFundingInputDb]): Future[
+    Vector[ScriptSignatureParams[InputInfo]]] = {
+    val outPoints =
+      fundingInputs.filter(_.isInitiator == dlcDb.isInitiator).map(_.outPoint)
 
-    val offerInputs =
-      fundingInputs.filter(_.isInitiator).map(_.toOutputReference)
-    val acceptInputs =
-      fundingInputs.filterNot(_.isInitiator).map(_.toOutputReference)
+    listUtxos(outPoints).map(_.map(info => info.toUTXOInfo(keyManager)))
```
Comment (benthecarman) [inline wallet/src/main/scala/org/bitcoins/wallet/DLCWallet.scala]: nit: 
> info => info.toUTXOInfo(keyManager)

_.toUTXOInfo(keyManager)

File: dlc/src/main/scala/org/bitcoins/dlc/sign/DLCTxSigner.scala [hunk 25]
```diff
@@ -0,0 +1,325 @@
...
+    }
+  }
+
+  def signRefundTx(remoteSig: PartialSignature): Future[Transaction] = {
+    for {
+      unsignedPSBT <- createPartiallySignedRefundTx()
+      psbt = unsignedPSBT.addSignature(remoteSig, inputIndex = 0)
+
+      refundTxT = psbt.finalizePSBT.flatMap(_.extractTransactionAndValidate)
+      refuntTx <- Future.fromTry(refundTxT)
```
Comment (benthecarman) [inline dlc/src/main/scala/org/bitcoins/dlc/sign/DLCTxSigner.scala]: > refuntTx

refundTx

File: wallet/src/main/scala/org/bitcoins/wallet/DLCWallet.scala [hunk 26]
```diff
@@ -362,53 +371,30 @@ abstract class DLCWallet extends Wallet {
   }
 
   def verifyCETSigs(sign: DLCSign): Future[Boolean] = {
```
Comment (benthecarman) [inline wallet/src/main/scala/org/bitcoins/wallet/DLCWallet.scala:366]: Now that we have a dedicated verifier, these functions should probably be moved to there

**Review by benthecarman — COMMENTED** [review]
Logic looks good since most is copy pasta from before, just a couple nits

File: wallet/src/main/scala/org/bitcoins/wallet/DLCWallet.scala (hunk 26 above)
Comment (nkohen) [reply wallet/src/main/scala/org/bitcoins/wallet/DLCWallet.scala:366]: So these are actually methods that call db functions and then call verifier functions so I think it is fine as is

**Review by benthecarman — APPROVED** [review]

**Pull request merged**